### PR TITLE
Solved issue #92

### PR DESCRIPTION
**Pull request by aalonzov**
Moved ShowDialog implementation from platform-specific interface to portable Window class. Marked old methods as obsolete.

### CODE REVIEW

**Comment by grokys** [comment]
Thanks for the PR @MrDaedra! I think that issue was a very old issue, and things have probably changed since, though I think some of the behavior could definitely be made x-plat. @kekekeks may have some thoughts on this.

**Comment by MrDaedra** [comment]
Yup, I know it's too old, but it looks reasonable and is still marked as pre-beta so you should deal with it soon. I found it trying to implement #624 and decided to solve it first. If it's unnecessary, let me do something else. 
Thanks.

**Comment by grokys** [comment]
I think this is good, except that `IWindowImpl.ShowDialog` shouldn't be made obsolete - call this from `Window.ShowDialog` instead of `Show()` in case the windowing subsystem needs to do anything special for a modal window.

File: src/Avalonia.Controls/Window.cs [hunk 1]
```diff
@@ -282,6 +284,32 @@ public Task<TResult> ShowDialog<TResult>()
                 return result.Task;
             }
         }
+        
+        private IDisposable GetModal()
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: This should probably be called something like `ShowModal` as its actually showing the modal dialog, not just getting it.

File: src/Avalonia.Controls/Window.cs [hunk 2]
```diff
@@ -307,7 +335,7 @@ protected override Size MeasureOverride(Size availableSize)
             var sizeToContent = SizeToContent;
             var size = ClientSize;
             var desired = base.MeasureOverride(availableSize.Constrain(_maxPlatformClientSize));
-
+
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Could you undo the whitespace changes?

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 3]
```diff
@@ -137,7 +137,7 @@ public Size MaxClientSize
         public WindowState WindowState
         {
             get
-            {
+            {
```
Comment (grokys) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: Whitespace

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 4]
```diff
@@ -176,7 +176,7 @@ public void Activate()
         }
 
         public IPopupImpl CreatePopup()
-        {
+        {
```
Comment (grokys) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: Whitespace

File: src/Avalonia.Controls/Window.cs [hunk 5]
```diff
@@ -348,7 +376,7 @@ private void EnsureInitialized()
                 init.BeginInit();
                 init.EndInit();
             }
-        }
+        }
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Whitespace

File: src/Avalonia.Controls/Window.cs [hunk 6]
```diff
@@ -282,6 +284,32 @@ public Task<TResult> ShowDialog<TResult>()
...
+            {
+                if (window.IsActive)
+                {
+                    activated = window;
+                }
+
+                window.IsEnabled = false;
+            }
+
+            PlatformImpl.Show();
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: We should still be calling `PlatformImpl.ShowDialog` here as setting up a modal dialog may require special steps in the windowing subsystem.

**Comment by kekekeks** [comment]
Calling `gtk_window_set_modal` is mandatory at least on Linux, since it sets  `_NET_WM_STATE_MODAL` X11 atom on the window. Therefore we can't avoid platform-specific implementation

**Comment by MrDaedra** [comment]
So does it mean that pull request and issue should be closed?

File: src/Avalonia.Controls/Window.cs [hunk 7]
```diff
@@ -45,7 +47,7 @@ public enum SizeToContent
     /// </summary>
     public class Window : TopLevel, IStyleable, IFocusScope, ILayoutRoot, INameScope
     {
-        private static IList<Window> s_windows = new List<Window>();
+        private static IList<Window> s_windows = new List<Window>();
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Could you undo these whitespace changes?

File: src/Avalonia.Controls/Window.cs [hunk 8]
```diff
@@ -89,7 +91,7 @@ static Window()
             TitleProperty.Changed.AddClassHandler<Window>((s, e) => s.PlatformImpl.SetTitle((string)e.NewValue));
             HasSystemDecorationsProperty.Changed.AddClassHandler<Window>(
                 (s, e) => s.PlatformImpl.SetSystemDecorations((bool) e.NewValue));
-
+
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Whitespace

File: src/Avalonia.Controls/Window.cs [hunk 9]
```diff
@@ -180,7 +182,7 @@ public WindowIcon Icon
         Size ILayoutRoot.MaxClientSize => _maxPlatformClientSize;
 
         /// <inheritdoc/>
-        Type IStyleable.StyleKey => typeof(Window);
+        Type IStyleable.StyleKey => typeof(Window);
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Whitespace

File: src/Avalonia.Controls/Window.cs [hunk 10]
```diff
@@ -238,7 +240,7 @@ public void Show()
                 PlatformImpl.Show();
             }
         }
-
+
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Whitespace

File: src/Avalonia.Controls/Window.cs [hunk 11]
```diff
@@ -263,14 +265,14 @@ public Task<TResult> ShowDialog<TResult>()
         {
             s_windows.Add(this);
 
-            EnsureInitialized();
-            LayoutManager.Instance.ExecuteInitialLayoutPass(this);
+            EnsureInitialized();
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Whitespace

File: src/Avalonia.Controls/Window.cs [hunk 12]
```diff
@@ -263,14 +265,14 @@ public Task<TResult> ShowDialog<TResult>()
         {
             s_windows.Add(this);
 
-            EnsureInitialized();
-            LayoutManager.Instance.ExecuteInitialLayoutPass(this);
+            EnsureInitialized();            
+            LayoutManager.Instance.ExecuteInitialLayoutPass(this);
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Whitespace

File: src/Avalonia.Controls/Window.cs [hunk 13]
```diff
@@ -263,14 +265,14 @@ public Task<TResult> ShowDialog<TResult>()
...
+            EnsureInitialized();            
+            LayoutManager.Instance.ExecuteInitialLayoutPass(this);                        
 
             using (BeginAutoSizing())
             {
-                var modal = PlatformImpl.ShowDialog();
+                var modal = GetModal();
                 var result = new TaskCompletionSource<TResult>();
-
+
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Whitespace

**Comment by grokys** [comment]
No, I think there you should continue with this PR. The main thing that needs to change is that this code:

https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WindowImpl.cs#L347-L358
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WindowImpl.cs#L362-L370

Should be moved out of the platform-specific libraries into `Window`. You have already done that, the things that need changing now are:

- Call `PlatformImpl.ShowDialog` here https://github.com/AvaloniaUI/Avalonia/pull/877/files#diff-fa616233ba8817223127ae63631cea21R302
- Remove the code you moved into `Window` from `IWindowImpl.ShowDialog` ([these](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WindowImpl.cs#L347-L358) [lines](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WindowImpl.cs#L362-L370))
- Remove the `[Obsolete]` attribute from `ShowDialog`
- Address the review comments I've left (mostly unnecessary whitespace changes)

**Comment by kekekeks** [comment]
IDisposable returned from `PlatformImpl.ShowDialog` is being ignored.

**Comment by grokys** [comment]
Thanks for the PR - sorry it took so long to get merged. I had to fix a couple of issues.

**Pull request merged**